iceoryx2/src/node/mod.rs

```diff
@@ -777,18 +796,27 @@ impl RegisteredServices {
         service_id: &ServiceId,
         mut cleanup_call: F,
     ) {
-        let mut data = self.data.lock().unwrap();
-        if let Some(entry) = data.get_mut(service_id) {
+        let mut guard = self.mutex().lock().expect("Failed to lock mutex");
+
+        if let Some(entry) = guard.get_mut(service_id) {
             entry.1 -= 1;
             if entry.1 == 0 {
-                cleanup_call(entry.0);
-                data.remove(service_id);
+                let handle = entry.0;
+                guard.remove(service_id);
+                drop(guard);
+                cleanup_call(handle);
```

Comment:
We need to be extremely careful here, not to introduce cleanup races. @elfenpiff can you please check if the changed order is problematic?